File: java/src/java/com/google/mr4c/nativec/ExternalAlgorithmSerializer.java
/**
  * Copyright 2014 Google Inc. All rights reserved.
  * 
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  * 
  *     http://www.apache.org/licenses/LICENSE-2.0
  * 
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

package com.google.mr4c.nativec;

import com.google.mr4c.algorithm.AlgorithmSchema;
import com.google.mr4c.serialize.AlgorithmSerializer;
import com.google.mr4c.serialize.SerializerFactory;

import java.io.IOException;
import java.io.StringReader;
import java.io.StringWriter;

public class ExternalAlgorithmSerializer {

	private AlgorithmSerializer m_serializer;
	private ExternalFactory m_factory;

	public ExternalAlgorithmSerializer(
		SerializerFactory serializerFactory,
		ExternalFactory factory
	) {
		m_serializer = serializerFactory.createAlgorithmSerializer();
		m_factory = factory;
	}

	public ExternalAlgorithm serializeAlgorithm(String name, AlgorithmSchema algo) throws IOException {
		ExternalAlgorithm extAlgorithm = m_factory.newAlgorithm(name);
		StringWriter writer = new StringWriter();
		m_serializer.serializeAlgorithmSchema(algo,writer);
		extAlgorithm.setSerializedAlgorithm(writer.toString());
		return extAlgorithm;
	}

	public AlgorithmSchema deserializeAlgorithm(ExternalAlgorithm extAlgorithm) throws IOException {
		AlgorithmSchema algo = new AlgorithmSchema();
		String serializedAlgo = extAlgorithm.getSerializedAlgorithm();
		StringReader reader = new StringReader(serializedAlgo);
		return m_serializer.deserializeAlgorithmSchema(reader);
	}

}

File: native/src/cpp/impl/external/ExternalAlgorithmSerializer.cpp
/**
  * Copyright 2014 Google Inc. All rights reserved.
  * 
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  * 
  *     http://www.apache.org/licenses/LICENSE-2.0
  * 
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

#include <log4cxx/logger.h>

#include "algorithm/algorithm_api.h"
#include "external/external_api.h"
#include "serialize/serialize_api.h"
#include "util/util_api.h"

using log4cxx::LoggerPtr;

namespace MR4C {

class ExternalAlgorithmSerializerImpl {

	friend class ExternalAlgorithmSerializer;

	private:
		
		LoggerPtr m_logger;
		AlgorithmSerializer* m_serializer;

		ExternalAlgorithmSerializerImpl(const SerializerFactory& factory) {
			
			m_logger = MR4CLogging::getLogger("external.ExternalAlgorithmSerializer");
			m_serializer = factory.createAlgorithmSerializer();
		}

		~ExternalAlgorithmSerializerImpl() {
			delete m_serializer;
		}

		void serializeAlgorithm(ExternalAlgorithm* extAlgo, const Algorithm& algo) const {
			std::string serAlgo = m_serializer->serializeAlgorithm(algo);
			extAlgo->setSerializedAlgorithm(serAlgo.c_str());
		}

		void deserializeAlgorithm(const ExternalAlgorithm& extAlgorithm, Algorithm& algo) const {
			m_serializer->deserializeAlgorithm(extAlgorithm.getSerializedAlgorithm(), algo);
		}

};

ExternalAlgorithmSerializer::ExternalAlgorithmSerializer(const SerializerFactory& factory) {
	m_impl = new ExternalAlgorithmSerializerImpl(factory);
}

ExternalAlgorithmSerializer::~ExternalAlgorithmSerializer() {
	delete m_impl;
}

void ExternalAlgorithmSerializer::serializeAlgorithm(ExternalAlgorithm* extAlgo, const Algorithm& algo) const {
	m_impl->serializeAlgorithm(extAlgo, algo);
}

void ExternalAlgorithmSerializer::deserializeAlgorithm(const ExternalAlgorithm& extAlgo, Algorithm& algo) const {
	m_impl->deserializeAlgorithm(extAlgo, algo);
}

}
